src/utils.js

```diff
@@ -78,6 +78,13 @@ const upload = async (opts) => {
   const response = await fetch(endpoint, { method: 'POST', headers, body });
   const uri = await response.text();
 
+  if (!uri) {
+    console.log(response);
```

Comment:
I would not log here and would throw the error message

Reply:
@DavidGOrtega, the issue is that there is no error message, but just an allegedly successful response with an empty body. That's why I chose to log the whole response object: it may contain precious information that might be hard to obtain on an intermttent bug like this.

Would you like to integrate it into the error message or just extract some properties?

Reply:
We should not log using console.log we should use a logging instead.
The error should bubble the issue to a better handler.
This is a technical debt that we have to solve with #22 
 - no console.log
 - no log that can not be display by debug, info, error levels...
 - prepare everything for json outputs choosing the right std (another log would break the output)

Reply:
> the issue is that there is no error message, but just an allegedly successful response with an empty body. That's why I chose to log the whole response object: it may contain precious information that might be hard to obtain on a (probably) intermittent bug like this.

Let's advise that the request to the storage failed. I wonder if we could not also retry at least once 🤔

Reply:
> _We should not log using console.log we should use a logging instead._

Agreed! I used a plain log call because we haven't moved yet to [`winston`](https://github.com/winstonjs/winston) as you proposed.

> _The error should bubble the issue to a better handler._

Then, I guess that we should just `throw` an error with templated information and let the handler do the rest.

Reply:
> _Let's advise that the request to the storage failed._

If it fails _for real_, we already get a proper error, and this edge case would only trigger when we receive an empty reply. Wouldn't treating this as a request error be slightly misleading?

I would throw something more technically accurate like «the asset storage returned an unexpected response» even if it doesn't matter that much as long as we identify the message.

> _I wonder if we could not also retry at least once_

Certainy, though it might be better to assume that the backend is reliable and, once it's proven to be irremediably unreliable, consider this kind of workaround.

Reply:
> Certainy, though it might be better to assume that the backend is reliable and, once it's proven to be irremediably unreliable, consider this kind of workaround.

I would not say it's not reliable, my guess is that its overloaded as any other SAAS. I would put that defense in order to protect the workflow. Take into account that if the user is not setting any kind of model cache she would have to redo the whole training because the asset did not uploaded at the end of the workflow.

Reply:
> _I would not say it's not reliable, my guess is that its overloaded as any other SAAS._

With the current backend architecture, I struggle to find how a possible overload would produce an empty succesful reply. Anyhow, we can't discard that possibility.

> _I would put that defense in order to protect the workflow. Take into account that if the user is not setting any kind of model cache she would have to redo the whole training because the asset did not uploaded at the end of the workflow._

That's a really good point, though I don't know if this event is frequent enough to justify this kind of safeguard. I wouldn't be concerned at all by this if we could silently send the error through a telemetry system, but I suspect that #443 won't happen that often and it could be more beneficial in the long run to produce an error, ask users to report, and fix the issue whenever possible.

Reply:
> I struggle to find how a possible overload would produce an empty succesful reply.

Do you have the code finally?

Reply:
Yes

Reply:
😆 where is it?

Reply:
[Here](https://us-east-2.console.aws.amazon.com/lambda/home?region=us-east-2#/functions/cml-image-upload). Sorry for the opaque reply, I thought that you noticed the thread in Slack. 😆